csrc/scheduler/mma_utils.cpp

```diff
@@ -1229,6 +1229,158 @@ RolesMapOpt getTensorsRoles(Fusion* fusion) {
   return roles_map;
 }
 
+namespace {
+
+void addMMAOp(Fusion* fusion_, std::vector<MulSumAsMmaProps>& props) {
+  auto* init = IrBuilder::create<Val>(0.0);
+  for (auto prop : props) {
+    IrBuilder::create<MmaOp>(prop.out, prop.a, prop.b, init);
+  }
+}
+
+// Check the val (in) is the output of broadcast.
+// Then check the output of the broadcast is 3D (4D for bmm).
+bool hasValidBroadcastOp(TensorView* bcast_out) {
+  // First check the tensorsview is 3D (4D)
+  // and has one broadcast dim.
+  auto dims = bcast_out->domain()->nDims();
+  if (!((dims == 3 || dims == 4) &&
+        bcast_out->domain()->noBroadcasts().size() == dims - 1)) {
+    return false;
+  }
+
+  // Check if the definition is a broadcast op.
+  if (dynamic_cast<BroadcastOp*>(bcast_out->definition())) {
+    return true;
+  }
+
+  return false;
+}
+
+bool broadcastsAreValid(
+    TensorView* left,
+    TensorView* right,
+    unsigned int reduction_axis) {
+  if (!(hasValidBroadcastOp(left) && hasValidBroadcastOp(right))) {
+    return false;
+  }
+
+  auto bcast_l = dynamic_cast<BroadcastOp*>(left->definition());
+  auto bcast_r = dynamic_cast<BroadcastOp*>(right->definition());
+
+  // Ensure that only one dim is getting broadcast.
+  auto bcastFlags_l = bcast_l->getBroadcastDimFlags();
+  auto bcastFlags_r = bcast_r->getBroadcastDimFlags();
+  auto count_l = std::count(bcastFlags_l.begin(), bcastFlags_l.end(), true);
+  auto count_r = std::count(bcastFlags_r.begin(), bcastFlags_r.end(), true);
+  if ((count_l != 1) || (count_l != count_r)) {
+    return false;
+  }
+
+  // Also ensure that it's not the same dim for the two muls. that's
+  // getting broadcast.
+  auto idx_l = std::find(bcastFlags_l.begin(), bcastFlags_l.end(), true) -
+      bcastFlags_l.begin();
+  auto idx_r = std::find(bcastFlags_r.begin(), bcastFlags_r.end(), true) -
+      bcastFlags_r.begin();
+  if (idx_l == idx_r) {
+    return false;
+  }
+
+  // Also ensure that the reduction dim is not either of the broadcast dim.
+  if (reduction_axis == idx_l || reduction_axis == idx_r) {
+    return false;
+  }
+
+  // Check different dimensions are the broadcast dims.
+  return true;
+}
+
+TensorView* getTensorviewPriorToCast(TensorView* in) {
+  if (in->getDataType() == DataType::BFloat16 ||
+      in->getDataType() == DataType::Half) {
+    return in;
+  }
+  TensorView* ret = nullptr;
+  if (in->getDataType() == DataType::Float) {
+    // get the producer.
+    auto* cast = in->definition();
+    auto* uCastOp = dynamic_cast<UnaryOp*>(cast);
+    if (uCastOp && uCastOp->getUnaryOpType() == UnaryOpType::Cast) {
+      auto* aTV = static_cast<TensorView*>(uCastOp->in());
+      NVF_CHECK(
+          (aTV->getDataType() == DataType::BFloat16 ||
+           aTV->getDataType() == DataType::Half),
+          "The output of the cast must be half or BF16");
+      ret = aTV;
+    }
+  }
+  return ret;
```

Comment:
So this function is expected to return nullptr if the input is a `Double` tensor. Is that right?

Reply:
Removing type checks here.